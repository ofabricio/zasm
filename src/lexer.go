package main

import (
	"io"
	"io/ioutil"
	"unicode"
	"unicode/utf8"
)

func Scan(r io.Reader) []*Token {
	lex := &Lexer{}
	return lex.Scan(r)
}

func (t *Lexer) Scan(r io.Reader) []*Token {

	data, _ := ioutil.ReadAll(r)

	var tokens []*Token

	row, col := 1, 1
	for len(data) > 0 {

		size, typ := t.tokenize(data)

		chars := utf8.RuneCount(data[:size])

		col += chars

		if typ == "NL" {
			row++
			col = 1
			typ = ""
		}

		if typ != "" {
			t := &Token{Text: string(data[:size]), Type: typ, Row: row, Col: col - chars}
			tokens = append(tokens, t)
		}

		data = data[size:]
	}

	return tokens
}

func (t *Lexer) tokenize(data []byte) (int, string) {
	r, size := utf8.DecodeRune(data)
	if r == '\n' {
		return size, "NL"
	}
	if unicode.IsSpace(r) {
		return size, ""
	}
	if r == ';' {
		return scan(data, Until('\n')), ""
	}
	if r == '\'' {
		return size + scan(data[size:], Until('\'')) + size, "STRING"
	}
	if IsAlphaNum_(r) {
		return scan(data, IsAlphaNum_), "WORD"
	}
	if IsTwinSymbol(r) {
		return scan(data, While(r)), "SYMBOL"
	}
	if IsSymbol(r) {
		return size, "SYMBOL"
	}
	return size, "UNKNOWN"
}

type Lexer struct {
}

type Token struct {
	Text string
	Type string
	Row  int
	Col  int
}

func scan(data []byte, cond MatchFunc) (adv int) {
	for {
		r, size := utf8.DecodeRune(data)
		if cond(r) {
			adv += size
		} else {
			break
		}
		data = data[size:]
	}
	return
}

func IsAlphaNum_(r rune) bool {
	return unicode.In(r, unicode.Letter, unicode.Number) || r == '_'
}

func IsSymbol(r rune) bool {
	return unicode.In(r, unicode.Symbol, unicode.Punct)
}

func IsTwinSymbol(r rune) bool {
	return r == '=' || r == '<' || r == '>' || r == '$'
}

func While(r rune) MatchFunc {
	return func(ru rune) bool {
		return ru == r
	}
}

func Until(r rune) MatchFunc {
	return func(ru rune) bool {
		return ru != r
	}
}

type MatchFunc func(rune) bool
