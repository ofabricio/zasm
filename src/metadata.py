"""
Tokens in Lexer:
    ('Type', 'token')
Tokens in Parser:
    ('Type', 'token', ...)
"""

registers = {
    'eax': {'Mod': 3, 'REG': 0, 'RM': 0, 'bits': 32},
    'ecx': {'Mod': 3, 'REG': 1, 'RM': 1, 'bits': 32},
    'edx': {'Mod': 3, 'REG': 2, 'RM': 2, 'bits': 32},
    'ebx': {'Mod': 3, 'REG': 3, 'RM': 3, 'bits': 32},
    'esp': {'Mod': 3, 'REG': 4, 'RM': 4, 'bits': 32},
    'ebp': {'Mod': 3, 'REG': 5, 'RM': 5, 'bits': 32},
    'esi': {'Mod': 3, 'REG': 6, 'RM': 6, 'bits': 32},
    'edi': {'Mod': 3, 'REG': 7, 'RM': 7, 'bits': 32},
}

instructions = {
    'mov': {'operands': 2},
    'add': {'operands': 2},
}

data_def = {
    'db',
    'dw',
}

symbol_table = {
    '(': 'LPar',
    ')': 'RPar',
    '[': 'LBra',
    ']': 'RBra',
    '{': 'LCur',
    '}': 'RCur',

    '+': 'AddOp',
    '-': 'SubOp',
    '*': 'MulOp',
    '/': 'DivOp',

    '@': 'At',
}

"""
Reg = 'Reg'
Inst = 'Inst'
Ident = 'Ident'
Symbol = 'Symbol'
Comment = 'Comment'
Program = 'Program'  # Parser

NL = 'NL'
Whitespace = 'Whitespace'

LPar = 'LPar'
RPar = 'RPar'
LBra = 'LBra'
RBra = 'RBra'
LCur = 'LCur'
RCur = 'RCur'

Expr = 'Expr'   # Parser
BinOp = 'BinOp'  # Parser
AddOp = 'AddOp'
SubOp = 'SubOp'
MulOp = 'MulOp'
DivOp = 'DivOp'

EOF = 'EOF'  # Parser
"""
