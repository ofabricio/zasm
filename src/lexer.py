import re
from metadata import registers, instructions, data_def, symbol_table


def tokenize(input):
    tokens = []

    pos = 0
    while len(input := input[pos:]) > 0:

        if pos := match(input, re_nl):
            type = 'NL'

        elif pos := match(input, re_whitespace):
            type = 'Whitespace'

        elif pos := match(input, re_comment):
            type = 'Comment'

        elif pos := match(input, re_alphanum):
            type = 'Ident'

        elif pos := match(input, re_symbols):
            type = 'Symbol'

        else:
            raise Exception(f'Unexpected token {input[0]}')

        if type in ['Whitespace', 'Comment']:
            continue

        token = input[:pos]
        tokens.append(typify(type, token))

    return tokens


def match(str, pattern):
    if m := re.match(pattern, str):
        return m.span(0)[1]
    return None


def typify(type, token):
    if token in symbol_table:
        type = symbol_table[token]
    elif token in registers:
        type = 'Reg'
    elif token in data_def:
        type = 'DD'
    elif token in instructions:
        type = 'Inst'
    elif match(token, re_num):
        type = 'Num'
    return type, token


re_nl = re.compile(r'\n')
re_whitespace = re.compile(r'\s')
re_alphanum = re.compile(r'[0-9a-zA-Z_]+')
re_comment = re.compile(r'\;.*')
re_symbols = re.compile(r'[()\[\]{}|&?*+\-<>$,.@#=]')
re_num = re.compile(r'^(([0-9A-F][0-9A-F])+|[0-9]+d)$')
