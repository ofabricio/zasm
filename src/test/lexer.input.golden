
val_to_hex_str
    pusha
    mov di HEX_OUT + 05
    mov cx 04
    > mov ax bx
    and al 0F
    cmp al 09
    jle >           ; jump to "add al 30" below.
    add al 07
    > add al 30
    mov [di] al
    dec di
    shr bx 04
    dec cx
    jnz <<          ; jump to "mov ax bx" above. A single '<' would jump to "add al 30".
    popa
    ret

HEX_OUT db '0x0000' 00

; ----

global_label
    mov eax ebx
.loc_label
    dec eax
    jnz .loc_label
    jz global_label.loc_label

; ----

    db 00
    dw 01 02 03 04
    db 'hello' 0A 00
    dd ?

    times 0200 - ($ - $$) db 00

; ----

    00
    arr 01 00 02 00 03 00

; Is the same as:

        db 00
    arr dw 01 02 03

; And:

    mov eax ebx
    90
    mov ecx ebx

; Is the same as:

    mov eax ebx
    nop
    mov ecx ebx

; ----

@org 7C00
@align 04
@include 'example.asm'

@struct Point {
    x db 00
    y db 00
}

@print {
}

@define VALUE 0123

@if VERSION == 0
    nop
@elseif VERSION == 1
    ; code...
@else
    ; code...
@end

@ifndef __PRINT__
@define __PRINT__
    ; code...
@end
@guard __PRINT__

@macro pushes a b {
    push a
    push b
}

    #pushes eax ebx
    call hello

; ----

    #example(eax, ebx, ecx, out r)
    mov eax r
    ret

@fn example (keep eax, ebx, stack v, out r) {
    mov eax ebx
    add eax v
    mov r eax
}

; ----

    sub  esp 04         ; "out" keyword reserves space for the returning value.
    push ecx            ; "stack" keyword uses the stack.
    call example
    add  esp 08
    mov  eax [esp-04]   ; [esp-04] is r from "out r"
    ret

example 
    push eax            ; "keep" keyword prevents trashing a register.
    mov  eax ebx
    add  eax [esp+04]   ; [esp+04] is v from "stack v"
    mov  [esp+08] eax   ; [esp+08] is r from "out r"
    pop  eax
    ret

; ----

@template 'GDT' code {
    Base        0
    Limit       0
    G           0
}
